### PR TITLE
Add network map WD-2133

### PR DESCRIPTION
**Pull request by edlerd**
Added a network map 
* with all the networks and all the instances in one image.
* with edges between instances that are in a network

As suggested in #151

### CODE REVIEW

**Comment by lorumic** [comment]
I think you should give full width to the cytoscape container. So that we avoid this:

image

**Comment by edlerd** [comment]
Changed some things
* using full screen size
* using an elastic algorithm to place the nodes on the map
* using standard vanilla colors

**Comment by Sensei-CHO** [comment]
Hi! glad you thought of my idea! I was able to recover some old stuff about network topology on LXD.

image
image
image

All of this were made with python pyvis. if I can bring a new idea based on my work, it could be great to have more information on instances and network by dragging the mouse on the plot, like in the screenshots.

Feel free to ask any questions or help. Thanks a lot!

**Comment by al-bakalova** [comment]
**Network table:**
- Please align the "Managed" column to the left
- Please align the IPV4 column to the right
- Please align "Used by" to the right
- Change button label to "See map"

**Network map:**
- Looks good as a first try. Let's keep it this way and come back to this once we have design resources to better understand the user stories behind

Ideas for future improvements:
- Add an icon for the legend and only display the legend on demand
- Not use green or red for the map elements since these colours usually mean success or error. I would use another colour for the running instances.
- It will be helpful to make the networks clickable and allow the user to browse another network. To do so, we need to introduce a "selected" state of the map elements
- Interactive maps usually have lots of accessibility issues. Let's thick of an alternative to display the information - maybe add a side panel with details.
- We should try to prevent the instance labels from overlapping. If not possible, maybe we can still consider a white background for the text.

**Comment by edlerd** [comment]
Added the white background for the labels on the map and adjusted the placement of the table columns.

**Comment by al-bakalova** [comment]
All changes look good to me! Thank you

File: static/sass/_network_map.scss [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#network-map {
```
Comment (yurii-vasyliev) [inline static/sass/_network_map.scss:1]: Could we assign a class and style by it? It's not a good practice to style by ID.

File: static/sass/_network_map.scss [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    .running {
+      background: #d9d9d9;
+    }
+
+    .other {
+      background: #0e8420;
+    }
+  }
+}
```
Comment (yurii-vasyliev) [inline static/sass/_network_map.scss:41]: If we replace hard-coded colors with sass variables, that would be great.

**Review by yurii-vasyliev — CHANGES_REQUESTED** [review]

File: static/js/pages/instances/DownloadButton.tsx [hunk 3]
```diff
@@ -48,19 +48,18 @@ const DownloadButton: FC<DownloadButtonProps> = ({
...
-        <Icon name="begin-downloading" alt="download" />
-      </Button>
-    </Tooltip>
+    <Button
+      appearance="base"
+      className="u-no-margin--bottom"
+      onClick={() => void handleDownload()}
+      title={`Download ${fileName}`}
+      aria-label={`Download ${fileName} file`}
+      has-icon
```
[the remark this answers is not in the document]
Comment (edlerd) [reply static/js/pages/instances/DownloadButton.tsx:57]: This change shouldn't be here, two commits sneaked into this one from #210 , I'll remove them here and address the issue over there.

File: static/js/pages/instances/DownloadButton.tsx [hunk 4]
```diff
@@ -48,19 +48,18 @@ const DownloadButton: FC<DownloadButtonProps> = ({
...
+      is-dense
```
[the remark this answers is not in the document]
Comment (edlerd) [reply static/js/pages/instances/DownloadButton.tsx:58]: This change shouldn't be here, two commits sneaked into this one from https://github.com/canonical/lxd-ui/pull/210 , I'll remove them here and address the issue over there.

File: static/js/pages/instances/actions/OpenTerminalBtn.tsx [hunk 5]
```diff
@@ -21,20 +21,16 @@ const OpenTerminalBtn: FC<Props> = ({
...
-        dense
-        hasIcon
-        onClick={handleOpen}
-        disabled={isDisabled}
-      >
-        <i className="p-icon--code" />
-      </Button>
-    </Tooltip>
+      <i className="p-icon--code" />
+    </Button>
```
[the remark this answers is not in the document]
Comment (edlerd) [reply static/js/pages/instances/actions/OpenTerminalBtn.tsx:33]: This change shouldn't be here, two commits sneaked into this one from https://github.com/canonical/lxd-ui/pull/210 , I'll remove them here and address the issue over there.

File: static/js/pages/instances/actions/OpenVgaBtn.tsx [hunk 6]
```diff
@@ -29,19 +29,18 @@ const OpenVgaBtn: FC<Props> = ({
...
+      dense
+      hasIcon
+      onClick={handleOpen}
+      disabled={
+        instance.status !== "Running" || instance.type !== "virtual-machine"
+      }
+      title={getTitle()}
+    >
+      <i className="p-icon--canvas" />
+    </Button>
```
[the remark this answers is not in the document]
Comment (edlerd) [reply static/js/pages/instances/actions/OpenVgaBtn.tsx:43]: This change shouldn't be here, two commits sneaked into this one from https://github.com/canonical/lxd-ui/pull/210 , I'll remove them here and address the issue over there.

File: static/js/pages/networks/NetworkMap.tsx [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            lineColor: "#D9D9D9",
+          },
+        });
+      })
+  );
+
+  const elements = [...networkNodes, ...instanceNodes, ...edges];
+
+  return (
+    <>
```
[the remark this answers is not in the document]
Comment (edlerd) [reply static/js/pages/networks/NetworkMap.tsx:133]: true dat.

File: static/sass/_instance_detail_panel.scss [hunk 8]
```diff
@@ -5,11 +5,14 @@
   }
 
   .p-instance-detail-panel--header {
+    background: white;
```
[the remark this answers is not in the document]
Comment (edlerd) [reply static/sass/_instance_detail_panel.scss:8]: This change shouldn't be here, two commits sneaked into this one from https://github.com/canonical/lxd-ui/pull/210 , I'll remove them here and address the issue over there.

File: static/sass/_network_map.scss (hunk 1 above)
Comment (edlerd) [reply static/sass/_network_map.scss:1]: I needed the id in JS, but we can have and ID and a class.

File: static/sass/_instance_detail_panel.scss (hunk 8 above)
[the remark this answers is not in the document]
Comment (edlerd) [reply static/sass/_instance_detail_panel.scss]: Actually this was removed in the other branch.

**Review by yurii-vasyliev — APPROVED** [review]

**Pull request merged**